### PR TITLE
Fix bug: if construct has name, it must appear in `end` statement

### PR DESCRIPTION
**Pull request by ZedThree**
Affects following block constructs:
- `associate`
- `do` (non-label)
- `if`
- `select case`
- `select type`
- `where`

This more strictly implements the constraints on the various blocks, which all have wording similar to the `do-construct`:

> C821 If the _do-stmt_ of a _block-do-construct_ specifies a _do-construct-name_, the corresponding
_end-do_ shall be an _end-do-stmt_ specifying the same _do-construct-name_. If the _do-stmt_ of a
_block-do-construct_ does not specify a _do-construct-name_, the corresponding _end-do_ shall not
specify a _do-construct-name_.

Basically, if you use `name: do`, you _must_ use `end do name`.

### CODE REVIEW

File: src/fparser/two/utils.py [hunk 1]
```diff
@@ -504,7 +504,8 @@ def match(startcls, subclasses, endcls, reader,
               enable_do_label_construct_hook=False,
               enable_if_construct_hook=False,
               enable_where_construct_hook=False,
-              strict_order=False):
+              strict_order=False,
+              strict_match_names=False):
```
Comment (ZedThree) [inline src/fparser/two/utils.py:508]: Alternative API: allow passing a string `"strict"` to `match_names`

**Comment by arporter** [comment]
Hi @ZedThree, just getting back up to speed with things. I just have a question about why we need a 'strict' option in addition to the `match_names` option. Once I understand that, I'll do a proper review (although I can warn you ahead of time that there are a lot of docstrings missing :-) ).

File: src/fparser/two/utils.py [hunk 2]
```diff
@@ -620,10 +623,14 @@ def match(startcls, subclasses, endcls, reader,
                         raise FortranSyntaxError(
                             reader, "Name '{0}' has no corresponding starting "
                             "name".format(end_name))
+                    elif strict_match_names and start_name and not end_name:
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/two/utils.py:626]: Yep, the name can be missed from `program`, `subroutine`, `function`, `module`, `interface` (possible others).

I think there might be a generic difference that `<block> name` constructs can skip `name`, while `name: <block>` constructs can't, but I haven't checked that fully. If so, there's almost certainly a more elegant way to handle this.

[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/two/utils.py:626]: I was thinking we could have a `NamedBlockBase` but then that would mean we would have e.g. both a `NamedAssociate` and an `Associate` class. Given that the name is optional and just an attribute I think we should only have one class for a given construct. Therefore, I think we'll stick with your approach (unless you have any better ideas :-) ).

**Comment by ZedThree** [comment]
Changes made. Adding the more explicit checks for the exception messages caught a bug: `else`/`case` statements aren't required to be named.

Also merged in master to fix conflicts

**Review by arporter — APPROVED** [review]
Thanks for making all those changes.
I've taken the liberty of fixing various lines that were too long (although I now find that there are a lot of those that are nothing to do with this PR).
I'll proceed to merge.

**Comment by arporter** [comment]
Actually, since I couldn't update the upstream branch (because it's in PlasmaFAIR), I merged it as is and created #346 to do the pycodestyle tidying.
